enclave.go

```diff
@@ -309,16 +349,177 @@ func (e *Enclave) Start() error {
 		return fmt.Errorf("%s: %w", errPrefix, err)
 	}
 
+	if !e.cfg.isScalingEnabled() {
+		return nil
+	}
+
+	// Check if we are the leader.
+	if !e.weAreLeader() {
+		elog.Println("Obtaining worker's hostname.")
+		worker := getSyncURL(getHostnameOrDie(), e.cfg.ExtPrivPort)
+		err = asWorker(e.setupWorkerPostSync, e.attester).registerWith(leader, worker)
+		if err != nil {
+			elog.Fatalf("Error syncing with leader: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// getSyncState returns the enclave's key synchronization state.
+func (e *Enclave) getSyncState() int {
+	e.Lock()
+	defer e.Unlock()
+	return e.syncState
+}
+
+// setSyncState sets the enclave's key synchronization state.
+func (e *Enclave) setSyncState(state int) {
+	e.Lock()
+	defer e.Unlock()
+	e.syncState = state
+}
+
+// weAreLeader figures out if the enclave is the leader or worker.
+func (e *Enclave) weAreLeader() (result bool) {
+	var (
+		err         error
+		ourNonce    nonce
+		weAreLeader = make(chan struct{}, 1)
+		areWeLeader = make(chan bool)
+		errChan     = make(chan error)
+		leader      = e.getLeader(pathLeader)
+	)
+	defer func() {
+		elog.Printf("We are leader: %v", result)
+		if result {
+			e.setSyncState(isLeader)
+		} else {
+			e.setSyncState(isWorker)
+		}
+	}()
+
+	ourNonce, err = newNonce()
+	if err != nil {
+		elog.Fatalf("Error creating new nonce: %v", err)
+	}
+
+	m := e.extPrivSrv.Handler.(*chi.Mux)
+	m.Get(pathLeader, getLeaderHandler(ourNonce, weAreLeader))
+	// Reset the handler as we no longer have a need for it.
+	defer m.Get(pathLeader,
+		func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(http.StatusGone)
+		},
+	)
+
+	timeout := time.NewTicker(10 * time.Second)
+	for {
+		go makeLeaderRequest(leader, ourNonce, areWeLeader, errChan)
+		select {
+		case <-e.stop:
+			return
+		case <-errChan:
+			elog.Println("Not yet able to talk to leader designation endpoint.")
+			time.Sleep(time.Second)
+			continue
+		case result = <-areWeLeader:
+			return
+		case <-weAreLeader:
+			e.setupLeader()
+			result = true
+			return
+		case <-timeout.C:
+			elog.Fatal("Timed out talking to leader designation endpoint.")
+		}
+	}
+}
+
+// setupWorkerPostSync performs necessary post-key synchronization tasks like
+// installing the given enclave keys and starting the heartbeat loop.
+func (e *Enclave) setupWorkerPostSync(keys *enclaveKeys) error {
+	e.keys.set(keys)
+	cert, err := tls.X509KeyPair(keys.NitridingCert, keys.NitridingKey)
+	if err != nil {
+		return err
+	}
+	e.httpsCert.set(&cert)
+
+	// Start our heartbeat.
+	worker := getSyncURL(getHostnameOrDie(), e.cfg.ExtPrivPort)
+	go e.workerHeartbeat(worker)
+
 	return nil
 }
 
+// setupLeader performs necessary setup tasks like starting the worker event
+// loop and installing leader-specific HTTP handlers.
+func (e *Enclave) setupLeader() {
```

Comment:
non-blocker: What's the expected method of handing leadership over if the leader node goes down? From my current understanding if the leader falls over as long as it properly restarts then when the worker contacts it during the heartbeat check and determines it's using the wrong key and resync's.

Wouldn't we run into a sync issue between the heartbeat recheck and the leader restarting? Seems that intermediate state is a concern but is short enough (looks to be once a minute) that probably not worth being concerned about. Is that aligned with your view?

Reply:
> Is that aligned with your view?

Yes. Heartbeats are cheap and we can increase their frequency if this turns out to be a concern.